File: assets/rule-templates/append-only-journal/java/src/main/java/Fakes.java
import java.util.ArrayList;
import java.util.List;

/**
 * Фейк журнала: только append и read — update и delete намеренно отсутствуют.
 *
 * <p>Фейк считает то, что проверяет инвариант: сколько записей добавлено
 * ({@link FakeJournal#appends()}) и что сейчас лежит в журнале
 * ({@link FakeJournal#read()}). Ни сети, ни сна.
 */
public final class Fakes {

    private Fakes() {
    }

    /**
     * Запись журнала: неизменяема по построению.
     *
     * @param reverses id исходной записи для компенсирующей; {@code null} — не компенсация
     */
    public record Entry(int id, String kind, long amount, Integer reverses) {
    }

    /** Журнал операций: записи только добавляются, id выдаётся по порядку. */
    public static final class FakeJournal {

        private final List<Entry> entries = new ArrayList<>();
        private int appends;

        /** Добавить запись; id растёт по порядку добавления. */
        public Entry append(String kind, long amount, Integer reverses) {
            appends++;
            Entry entry = new Entry(entries.size() + 1, kind, amount, reverses);
            entries.add(entry);
            return entry;
        }

        /**
         * Записи журнала в порядке добавления.
         *
         * <p>Фейк отдаёт внутренний список как есть — это осознанное упрощение:
         * неизменяемость журнала обеспечивает вызывающий, и именно её проверяет
         * инвариант.
         */
        public List<Entry> read() {
            return entries;
        }

        /** Сколько записей добавлено. */
        public int appends() {
            return appends;
        }
    }
}

File: assets/rule-templates/append-only-journal/python/fakes.py
"""Фейк журнала: только append и read — update и delete намеренно отсутствуют.

Фейк считает то, что проверяет инвариант: сколько записей добавлено
(`appends`) и что сейчас лежит в журнале (`read`). Ни сети, ни сна —
детерминированный прогон за миллисекунды.
"""


class FakeJournal:
    """Журнал операций: записи только добавляются, id выдаётся по порядку."""

    def __init__(self) -> None:
        self.appends = 0
        self._entries: list[dict] = []

    def append(self, kind: str, amount: int, reverses=None) -> dict:
        """Добавить запись; id растёт по порядку добавления."""
        self.appends += 1
        entry = {
            "id": len(self._entries) + 1,
            "kind": kind,
            "amount": amount,
            "reverses": reverses,
        }
        self._entries.append(entry)
        return entry

    def read(self) -> list[dict]:
        """Записи журнала в порядке добавления.

        Фейк отдаёт внутренний список как есть — это осознанное упрощение:
        неизменяемость журнала обеспечивает вызывающий, и именно её проверяет
        инвариант.
        """
        return self._entries

    def count(self) -> int:
        """Сколько записей в журнале."""
        return len(self._entries)
